src/v/kafka/protocol/tests/field_parser_test.cc

```diff
@@ -213,3 +228,241 @@ SEASTAR_THREAD_TEST_CASE(serde_flex_types) {
+            auto response = rg::gen<kafka::api_versions_response>::generate(
+              (i % 2) == 0);
+            auto serded = serde_kafka(response, kafka::api_version(3));
+            BOOST_CHECK_EQUAL(response, serded);
+            return ss::now();
+        }).get();
+    }
+    {
+        /// Test backward compatability support, given a newer struct, do not
+        /// encode / decode tags
+        auto response = rg::gen<kafka::api_versions_response>::generate(true);
+        auto serded = serde_kafka(response, kafka::api_version(1));
+        response.data.supported_features.clear();
+        response.data.finalized_features.clear();
+        response.data.finalized_features_epoch = -1; // default
+        BOOST_CHECK_EQUAL(response, serded);
+    }
+    {
+        /// Test that newer versions without tags don't encode them
+        auto response = rg::gen<kafka::api_versions_response>::generate(false);
+        auto serded = serde_kafka(response, kafka::api_version(3));
+        BOOST_CHECK_EQUAL(response, serded);
+        BOOST_CHECK(serded.data.finalized_features.empty());
+        BOOST_CHECK(serded.data.supported_features.empty());
+        BOOST_CHECK_EQUAL(serded.data.finalized_features_epoch, -1);
+    }
+    /// Measure a static expected size
+    kafka::api_versions_response response{
+      .data = kafka::api_versions_response_data{
+        .api_keys = rg::gen_array<kafka::api_versions_response_key>(5)}};
+    {
+        iobuf b;
+        kafka::response_writer writer(b);
+        response.encode(writer, kafka::api_version(3));
+        /// error_code = 2 bytes
+        /// throttle_time_ms = 4 bytes
+        /// api_keys =
+        ///    size = 5 encoded as uvarint = (1 byte)
+        ///    api_versions_response_key = 6 bytes + 1 zero byte for empty tags
+        ///    <- repeated 5 times = 35 bytes total
+        /// outer tags = 1 byte (to denote empty)
+        /// total = 2 + 4 + (1 + 35) + 1
+        BOOST_CHECK_EQUAL(43, b.size_bytes());
+    }
+    {
+        /// fill in a non default tag
+        response.data.finalized_features_epoch = 52;
+        iobuf b;
+        kafka::response_writer writer(b);
+        response.encode(writer, kafka::api_version(3));
+        /// Main difference is tag section will no longer be 1 byte
+        ///
+        /// size of tags = 1 byte uvarint encoded
+        /// encoding tag itself = 1 byte uvarint encoded
+        /// encoding of the value = 8 bytes encoded as int64_t
+        /// size of the value itself = 1 byte (sizeof(int64_t))
+        /// total = 11 for the tags section, delta of 10 bytes
+        BOOST_CHECK_EQUAL(53, b.size_bytes());
+    }
+}
+
+// The following hex data was generated by the go code pasted below. The idea is
+// to have our parsers test correctly formatted and serialized binary data
+// generated by franz-go
+static const unsigned char test_api_version_response_data[78] = {
```

Comment:
Could we gen the full type suite programmatically as input via a a Python script or bash script of sorts and have the cpp parse it

Reply:
In fact we could generate the entire test via some out of band script into a single .cc file

Reply:
Yes, but only 3-4 Kafka requests really define tags (its a feature not really used), and furthermore we don't have the latest up to date schemas so i think only 2 of those would produce generated code that uses tags.

But that idea is better suited to test parsing in general, tags or no tags, probably best suited for testing the flex PR, and I agree with the sentiment. Ideally we'd use `franz-go` to generate the dataset, as the `kafka-python` dependency is too old, doesn't support flex/compact types.

Reply:
File a ticket https://github.com/redpanda-data/redpanda/issues/5117